### ISSUE TITLE
pico controller position offset

### ISSUE THREAD

**Issue by DanielKinsman** [WiVRn/WiVRn#94]
Only just started with wivrn and it is great, much smoother experience doing xr dev in godot using wivrn instead of alvr + steamvr.

One thing I noticed though is that the pico 4 controllers are offset compared to where they should be, the video should show it. It starts in wivrn and then I pres the "home" button and it shows where they are according to the pico OS home screen.

https://github.com/user-attachments/assets/4fb53c8f-b41e-40ce-9577-92ff3df8a0a7

**Comment by xytovl** [comment]
Can you detail:
* is the pointing correct in the lobby?
* is the pointing correct when streaming?
* is the "grip" correct when streaming?

This would help understand if it is an issue with the graphical model, interaction profiles or something else.

**Comment by DanielKinsman** [comment]
* pointing is incorrect in lobby
* pointing is incorrect when streaming
* grip is incorrect when streaming

**Comment by xytovl** [comment]
During stream, raw data is forwarded. The only thing I can see is the reported interaction profile.

Try to replace https://github.com/Meumeu/WiVRn/blob/master/server/driver/wivrn_controller.cpp#L168 with `XRT_DEVICE_PICO4_CONTROLLER` and see if it makes any difference.
If it fixes the issue I'll have to forward it from client.

**Comment by DanielKinsman** [comment]
That makes the controllers inoperable in game. It also doesn't change anything in the lobby.

There's a small chance the controllers being inoperable could possibly be a godot thing though, so I'll test with some other software. Do you know if there is anything like xrgears / hello_xr that uses controllers?

**Comment by damaestro** [comment]
https://gitlab.freedesktop.org/monado/demos/openxr-simple-playground has controller input, but maybe not something that would help diagnose.

**Comment by xytovl** [comment]
For the lobby, it could be an issue with model or the way we use it.

For the stream we directly forward data for `/user/hand/{left,right}/input/{grip,aim}/pose`, I don't understand how this could be wrong.
If you build your application (as long as it is native openxr) for Pico directly, does it behave differently?

**Comment by DanielKinsman** [comment]
Yeah the native pico version of my game works fine as expected.

When streaming, both my game and [open saber](https://leandrodreamer.itch.io/open-saber/) (also godot based) suffer from the offset. When pointing at the horizon with zero roll axis the pointer is angled upwards a fair bit.

Testing with streaming openxr-simple-playground the rectangles (assuming they represent the controllers) do have the correct orientation, being in line with the controller grip, but there is no pointer to test.

I'll try with some other non-godot apps and get back to you.

**Comment by xytovl** [comment]
It seems that Godot doesn't use the Pico specific interaction profile: https://github.com/godotengine/godot/blob/master/modules/openxr/action_map/openxr_interaction_profile_metadata.cpp

Can you try apk from branch https://github.com/Meumeu/WiVRn/tree/test/pico-interaction-profile and see if it works and if it makes any difference in controller pose? CI will build it in https://github.com/Meumeu/WiVRn/actions/workflows/Build.yml?query=branch%3Atest%2Fpico-interaction-profile

**Comment by DanielKinsman** [comment]
I tried the pico release apk from that build against the standard wivrn-server from arch aur, controllers didn't work at all while streaming in my game. Let me know if I should be grabbing the server from there too.

Godot has an extension here https://github.com/godotengine/godot/blob/607b230ffe120b2757c56bd3d52a7a0d4e502cfe/modules/openxr/extensions/openxr_pico_controller_extension.cpp

They also use "action maps" and configure this stuff at runtime I think.
https://docs.godotengine.org/en/stable/tutorials/xr/xr_action_map.html
In my game I am using the default action map given by godot 4.2. https://gitlab.com/DanielKinsman/whos-a-good-boy/-/blob/main/gdproject/openxr_action_map.tres?ref_type=heads#L451

**Comment by xytovl** [comment]
I have tested on Quest and HTC Vive XR elite, in lobby models are not perfect, but the aim ray matches what is in the home screen both in lobby and when streaming.

If this issue is indeed specific to Pico I can not fix it myself.

Can you provide the application where the issue happens so that I can see if I can reproduce on my hardware?

**Comment by DanielKinsman** [comment]
https://copper-tunic.itch.io/whos-a-good-boy
Code is at https://gitlab.com/DanielKinsman/whos-a-good-boy

While the pointer is off somewhat, it is useable, but the hand position is what is really off. Also happens in open saber (linked earlier in this issue) and prob everything else godot based. I have taken some screenshots and video to show it but haven't pulled them off the headset yet and I'm not at the computer today.

I also haven't yet done the testing with non-godot based things yet, will get to that tomorrow and let you know how it goes. It's probably worth saving yourself any more trouble looking into this for now until I do that.

**Comment by DanielKinsman** [comment]
TL;DR I think this is because of wivrn / monado missing `XR_EXT_palm_pose`

It does appear to only be happening in godot.

https://github.com/user-attachments/assets/bcab3664-721d-41d5-92a8-76720b4bc80e

In case it matters, the dog one (first) was run natively while the others were run via steam proton experimental->opencomposite->openxr.

The most relevant info though is probably the verbose output from godot itself.

* [alvr.log](https://github.com/user-attachments/files/16406669/alvr.log)
* [wivrn.log](https://github.com/user-attachments/files/16406672/wivrn.log)

Of special note is this stuff only in the wirvn output:

```
LOG in xrCreateInstance: Selected devices
	Head: 'WiVRn HMD'
	Eyes: '<none>'
	Left: 'WiVRn HMD left hand controller'
	Right: 'WiVRn HMD right hand controller'
	Gamepad: '<none>'
	Hand-Tracking Left: 'WiVRn HMD left hand controller'
	Hand-Tracking Right: 'WiVRn HMD right hand controller'
```

```
OpenXR: IO path /interaction_profiles/bytedance/pico4_controller/user/hand/left/input/palm_ext/pose requires extension XR_EXT_palm_pose
OpenXR: IO path /interaction_profiles/bytedance/pico4_controller/user/hand/right/input/palm_ext/pose requires extension XR_EXT_palm_pose
```

**Comment by DanielKinsman** [comment]
Hmm then again, maybe not. I toggled it on in [monado's cmake](https://gitlab.freedesktop.org/monado/monado/-/blob/main/CMakeLists.txt?ref_type=heads#L361) and godot stopped complaining about the missing extension, but the behaviour was the same. Maybe that option is disabled by default for a reason though?

**Comment by xytovl** [comment]
Added palm_ext in https://github.com/WiVRn/WiVRn/tree/feat/palm_ext

**Comment by DanielKinsman** [comment]
I think after all that while the particular godot stuff I am using complains about the extension missing, it doesn't actually use it for locating the hands, everything is still based off the pointer location. But at least with the extension available it should fix it in the wivrn dashboard and I might be able to start using the extension in my projects to fix the problem.

**Comment by Mhowser** [comment]
> Added palm_ext in https://github.com/WiVRn/WiVRn/tree/feat/palm_ext

When can this be merged to master?

**Comment by xytovl** [comment]
The branch was kind of forgotten.
I have just rebased it but it seems I am unable to test: Quest 1 does not support the extension and HTC XR Elite always returns empty data.

**Comment by galister** [comment]
https://github.khronos.org/OpenXR-Inventory/extension_support.html#XR_EXT_palm_pose
only the Pico 4 Ultra is listed as supported. that might mean other Pico headsets as well, but likely no Meta/HTC

**Comment by Mhowser** [comment]
I rebuilt the client apk with https://github.com/WiVRn/WiVRn/commit/6608b1d833ec187f49fab7fc508fbe51fb5efb87 and the offset is still present on Pico 4 HMD
https://github.com/user-attachments/assets/bc5607c0-d960-42c7-96fd-efd8c262fceb

**Comment by xytovl** [comment]
In 33bde8be579ad303b74d836516d0ea3f648419d7 there is a new debug menu that can be enabled at compile time.
This menu can apply offsets for controller models in the lobby.

In order to fix the values on Pico, someone with access to the headset should:
edit build.gradle to add `-DWIVRN_CLIENT_DEBUG_MENU=ON`
build android client per docs/building.md
manually adjust the offsets
set the default offsets in `client/hardware.cpp`
make a pull request to integrate values.

**Comment by Mhowser** [comment]
I don't see an entry for Pico 4 controller offsets below this line https://github.com/WiVRn/WiVRn/blob/c7bcb561f07bb7642cbbf462cdecfa15c02ade43/client/hardware.cpp#L236 do we need to add it in or is are the controller offsets somewhere else?

**Comment by Mhowser** [comment]
I built the client apk and found good values for the controller offsets:

XPos: 0.000 YPos: -0.030 Zpos: -0:040

XRot: -35.000 YRot: 0.000 ZRot: 0.000

however there is no way I can adjust the separation value of the controllers as seen in the video:

https://github.com/user-attachments/assets/548ab7e6-460e-4569-8b6c-1451d45d5d4e
